pyodide-build/pyodide_build/mkpkg.py

```diff
@@ -10,11 +10,37 @@
 import urllib.request
 import warnings
 from pathlib import Path
-from typing import Any, Literal
+from typing import Any, Literal, TypedDict
 
 from ruamel.yaml import YAML
 
 
+class URLDict(TypedDict):
```

Comment:
You don't need all arguments (total=False is the default), only the ones you access, but I went ahead and added them.

Reply:
Wow neat. We could use this in a bunch of places.

Reply:
Yes, you can fully type all your dict interfaces, which is really nice. 3.11 added Required/NotRequired so you can specify individual keys rather than making a complete or minimal TypedDict with total=True/False.

If you are building your own data, though, dataclass is the right way to go. Dicts should be used like this, for reading external APIs and files.

Reply:
(I could have also handled nested items here, but this was fine to check most of the usage)

Reply:
> dataclass is the right way to go. Dicts should be used like this, for reading external APIs and files.

Thanks for the tip. I switched micropip to using dataclasses for `Transaction` and `Wheel` and it was indeed a much better approach because then it became apparent that I could reorganize a lot of the functions into methods on these classes and the code clarity improved a lot.

Reply:
Ideally it would be nice to share more logic between `pyodide_build` and `micropip` since there is a significant amount of overlap.

Reply:
> Dicts should be used like this, for reading external APIs and files.

At least for meta.yaml we can probably switch to Pydantic to validate the spec, rather than doing our own custom dict validations.